compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -41,66 +41,62 @@ namespace mlir::iree_compiler {
 #define GEN_PASS_DEF_GPUMATERIALIZEHOSTENCODINGPASS
 #include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
 
-static bool hasIntrinsic(IREE::GPU::TargetAttr target,
-                         IREE::GPU::MMAIntrinsic intrinsic) {
-  for (IREE::GPU::MMAAttr mma : target.getWgp().getMma()) {
-    if (mma.getIntrinsic().getValue() == intrinsic) {
-      return true;
+static IREE::GPU::MMAAttr chooseIntrinsicMMAAttr(TypeRange eTypes,
+                                                 IREE::GPU::TargetWgpAttr wgp) {
+  IREE::GPU::MMAAttr candidateMma;
+  for (auto mma : wgp.getMma()) {
+    // Filter out intrinsics that don't match the element types of this matmul.
+    auto [et0, et1, et2] = mma.getABCElementTypes();
+    if (et0 != eTypes[0] || et1 != eTypes[1] || et2 != eTypes[2]) {
+      continue;
+    }
+    // If multiple intrinsics are available for the given element types, we have
+    // to make a choice. On CDNA3, there may be an intrinsic with larger M/N and
+    // smaller K, which would optimize power, and an intrinsic with larger K,
+    // which would optimize performance when power is not the bottleneck.
+    // Currently we just choose the intrinsic maximizing K, but that can be
+    // revisited later.
+    if (candidateMma && candidateMma.getKSize() > mma.getKSize()) {
+      continue;
```

Comment:
Do we ever enter this `if` since there's a `break` just after the only assignment to `candidateMma`?

Reply:
Great catch, thanks!